pkg/cmd/factory/factory_impl.go

```diff
@@ -0,0 +1,148 @@
+	"github.com/pingcap/ticdc/pkg/version"
+	pd "github.com/tikv/pd/client"
+	"go.etcd.io/etcd/clientv3"
+	etcdlogutil "go.etcd.io/etcd/pkg/logutil"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+)
+
+type factoryImpl struct {
+	clientGetter ClientGetter
+}
+
+// NewFactory creates a client build factory.
+func NewFactory(clientGetter ClientGetter) Factory {
+	if clientGetter == nil {
+		panic("attempt to instantiate factory with nil clientGetter")
+	}
+	f := &factoryImpl{
+		clientGetter: clientGetter,
+	}
+
+	return f
+}
+
+func (f *factoryImpl) ToTLSConfig() (*tls.Config, error) {
+	return f.clientGetter.ToTLSConfig()
+}
+
+func (f *factoryImpl) ToGRPCDialOption() (grpc.DialOption, error) {
+	return f.clientGetter.ToGRPCDialOption()
+}
+
+func (f *factoryImpl) GetPdAddr() string {
+	return f.clientGetter.GetPdAddr()
+}
+
+func (f *factoryImpl) GetCredential() *security.Credential {
+	return f.clientGetter.GetCredential()
+}
+
+func (f *factoryImpl) EtcdClient() (*kv.CDCEtcdClient, error) {
+	ctx := cmdconetxt.GetDefaultContext()
+
+	tlsConfig, err := f.ToTLSConfig()
+	if err != nil {
+		return nil, err
+	}
+	grpcTLSOption, err := f.ToGRPCDialOption()
+	if err != nil {
+		return nil, err
+	}
+
+	logConfig := etcdlogutil.DefaultZapLoggerConfig
+	logConfig.Level = zap.NewAtomicLevelAt(zapcore.ErrorLevel)
+	pdEndpoints := strings.Split(f.GetPdAddr(), ",")
+
+	etcdClient, err := clientv3.New(clientv3.Config{
+		Context:     ctx,
+		Endpoints:   pdEndpoints,
+		TLS:         tlsConfig,
+		LogConfig:   &logConfig,
+		DialTimeout: 30 * time.Second,
+		DialOptions: []grpc.DialOption{
+			grpcTLSOption,
+			grpc.WithBlock(),
+			grpc.WithConnectParams(grpc.ConnectParams{
+				Backoff: backoff.Config{
+					BaseDelay:  time.Second,
+					Multiplier: 1.1,
+					Jitter:     0.1,
+					MaxDelay:   3 * time.Second,
+				},
+				MinConnectTimeout: 3 * time.Second,
+			}),
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	client := kv.NewCDCEtcdClient(ctx, etcdClient)
+	return &client, nil
+}
+
+func (f factoryImpl) PdClient() (pd.Client, error) {
+	ctx := cmdconetxt.GetDefaultContext()
+
+	credential := f.GetCredential()
+	grpcTLSOption, err := f.ToGRPCDialOption()
+	if err != nil {
+		return nil, err
+	}
+
+	pdAddr := f.GetPdAddr()
+	pdEndpoints := strings.Split(pdAddr, ",")
+
+	pdClient, err := pd.NewClientWithContext(
+		ctx, pdEndpoints, credential.PDSecurityOption(),
+		pd.WithGRPCDialOptions(
+			grpcTLSOption,
+			grpc.WithBlock(),
+			grpc.WithConnectParams(grpc.ConnectParams{
+				Backoff: backoff.Config{
+					BaseDelay:  time.Second,
+					Multiplier: 1.1,
+					Jitter:     0.1,
+					MaxDelay:   3 * time.Second,
+				},
+				MinConnectTimeout: 3 * time.Second,
+			}),
+		))
+	if err != nil {
+		return nil, errors.Annotatef(err, "fail to open PD client, pd=\"%s\"", pdAddr)
+	}
+
+	err = version.CheckClusterVersion(ctx, pdClient, pdEndpoints[0], credential, true)
```

Comment:
Off topic: I think we need to check all pd endpoint and make sure they belong to the same cluster.

Reply:
Added a to-do item. https://github.com/pingcap/ticdc/pull/2341/commits/de0fae2585b99dd2cc666fef88172e46c93e92c9